include/unifex/detail/with_type_erased_tag_invoke.hpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/this.hpp>
+#include <unifex/detail/vtable.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+template <
+    typename Derived,
+    typename CPO,
+    bool NoExcept,
+    typename Sig>
+struct _with_type_erased_tag_invoke;
+
+template <
+    typename Derived,
+    typename CPO,
+    bool NoExcept,
+    typename Ret,
+    typename... Args>
+struct _with_type_erased_tag_invoke<
+    Derived,
+    CPO,
+    NoExcept,
+    Ret(Args...)> {
+  struct type {
+    friend Ret tag_invoke(base_cpo_t<CPO> cpo, replace_this_t<Args, Derived>... args)
+        noexcept(NoExcept) {
+      using cpo_t = base_cpo_t<CPO>;
+      static_assert(
+          !NoExcept || noexcept(extract_this<Args...>{}((decltype(args) &&) args...)));
+      auto&& t = extract_this<Args...>{}((decltype(args) &&) args...);
+      void* objPtr = get_object_address(t);
+      auto* fnPtr = get_vtable(t)->template get<CPO>();
+
+      // Sanity-check noexceptness of implementation.
```

Comment:
I like that the comment on the sanity check in the forwarding implementation includes the justification for the sanity check; consider copying it here.